In [1]:
import warnings
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import copy
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
warnings.filterwarnings('ignore')

train = pd.read_csv("cat-reloaded-data-science-competition-2-0\train.csv")
test = pd.read_csv("cat-reloaded-data-science-competition-2-0\test.csv")

In [2]:
train.head()

ID          Brand                                       VehicleModel  \
0   2624            NaN  I have the fastest model ever.. yes you got it...   
1  14392  I love TOYOTA                                                NaN   
2   9822        hyundai                                                 H1   
3  17123           opel                                              Corsa   
4  11133         subaru                                     I own Outback.   

  ManufacturingYear        Type  rating          color  \
0              2017  WOOW Sedan     0.0            NaN   
1               NaN    Minivan3     5.0       Metallic   
2              2009     MINIVAN     0.0       Metallic   
3               NaN   Hatchback     5.0      Jet Black   
4              2007   Hatchback     4.0  Sapphire Blue   

                                                Duty  fuel CylinderCount  \
0  I had to pay a substantial $1249 levy on my ne...   Gas           4.0   
1                                                  -   Gas       4.0 cyl   
2                                              1185$  Dies           4.0   
3                                                  -   NaN           4.0   
4                                                  -   Gas           4.0   

  type of gear capacity        Odometer #airbags  price  
0            R      2.5        75492 KM     1008  21012  
1            R      2.4       220000 KM      -96  11917  
2            R      2.5  TOTAL 50653 km  4 count  22433  
3         auto        0       230000 KM      0     5645  
4          NaN      2.5       131000 KM      4    10976

In [3]:
test.head()

ID          Brand                                       VehicleModel  \
0   5634            NaN  I have the fastest model ever.. yes you got it...   
1  16689          HONDA                                I own Civic Hybrid.   
2  13457     I love BMW                                          I own M3.   
3   4555        hyundai                                            ELANTRA   
4   9726  mercedes-benz                my model which is E 300 is the best   

  ManufacturingYear       Type  rating      color  \
0               NaN      SEDAN     4.0  Jet Black   
1             2012;      sedan     1.0   Sky Blue   
2              2006  hatchback     1.0        NaN   
3              3016      sedan     3.0        NaN   
4               NaN      Sedan     8.0        NaN   

                                                Duty             fuel  \
0  I had to pay a substantial $1017 levy on my ne...              NaN   
1                                                  =  Electric Hybrid   
2                                                  -             Dies   
3                                         2188 pound              Gas   
4                                               1017              NaN   

  CylinderCount type of gear   capacity         Odometer #airbags  
0           4.0         auto          2             1600  12 bags  
1           4.0         auto        1.5  TOTAL 141302 km       12  
2           4.0       manual  1.9 Turbo     136000  kilo        6  
3           4.0         gear        1.8            78000       10  
4           4.0            R          2  1600  kilometer       12

In [4]:
# I just did this function because in my case I turned all columns to string type and I had to figure out another way to check missing values

def is_nan(column):
    if column.dtype == 'object':
        return column.isna().sum() + column.isin(['', ' ', 'nan', 'NAN', 'NaN', 'na', 'NA', 'n/a', 'N/A', 'null', 'NULL']).sum()
    else:
        return column.isna().sum()

In [5]:
manipulated = train.copy()
for col in manipulated.columns[manipulated.dtypes == 'object']:
    manipulated[col] = manipulated[col].astype('str')

### Summary of train data


In [6]:
train.shape

(7000, 15)

In [7]:
train.describe()

ID       rating         price
count   7000.000000  6310.000000   7000.000000
mean    9022.128714     2.559588  14406.866429
std     5202.316292     1.905482  11747.617978
min        1.000000    -4.000000      3.000000
25%     4546.750000     1.000000   5018.000000
50%     9044.500000     2.000000  12544.000000
75%    13523.250000     4.000000  20071.000000
max    18013.000000    10.000000  49889.000000

In [8]:
train.describe(include=['O'])

Brand VehicleModel ManufacturingYear   Type      color  Duty  fuel  \
count     6294         6303              6305   6315       4876  7000  6304   
unique     185         1480               555     81         16  1173     7   
top     toyota        Prius              2012  sedan  Jet Black     -   Gas   
freq       497          113               429   1160       1239  2108  3272   

       CylinderCount type of gear capacity Odometer #airbags  
count           7000         6308     7000     7000     7000  
unique            28            7       87     4964      321  
top              4.0         auto        2     0 km       12  
freq            2638         2870     1261       79      830

In [9]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7000 entries, 0 to 6999
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ID                 7000 non-null   int64  
 1   Brand              6294 non-null   object 
 2   VehicleModel       6303 non-null   object 
 3   ManufacturingYear  6305 non-null   object 
 4   Type               6315 non-null   object 
 5   rating             6310 non-null   float64
 6   color              4876 non-null   object 
 7   Duty               7000 non-null   object 
 8   fuel               6304 non-null   object 
 9   CylinderCount      7000 non-null   object 
 10  type of gear       6308 non-null   object 
 11  capacity           7000 non-null   object 
 12  Odometer           7000 non-null   object 
 13  #airbags           7000 non-null   object 
 14  price              7000 non-null   int64  
dtypes: float64(1), int64(2), object(12)
memory usage: 820.4+ KB


In [10]:
train.columns[train.dtypes == 'object']

Index(['Brand', 'VehicleModel', 'ManufacturingYear', 'Type', 'color', 'Duty',
       'fuel', 'CylinderCount', 'type of gear', 'capacity', 'Odometer',
       '#airbags'],
      dtype='object')

In [11]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7000 entries, 0 to 6999
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ID                 7000 non-null   int64  
 1   Brand              6294 non-null   object 
 2   VehicleModel       6303 non-null   object 
 3   ManufacturingYear  6305 non-null   object 
 4   Type               6315 non-null   object 
 5   rating             6310 non-null   float64
 6   color              4876 non-null   object 
 7   Duty               7000 non-null   object 
 8   fuel               6304 non-null   object 
 9   CylinderCount      7000 non-null   object 
 10  type of gear       6308 non-null   object 
 11  capacity           7000 non-null   object 
 12  Odometer           7000 non-null   object 
 13  #airbags           7000 non-null   object 
 14  price              7000 non-null   int64  
dtypes: float64(1), int64(2), object(12)
memory usage: 820.4+ KB


In [12]:
train.isnull().sum()

ID                      0
Brand                 706
VehicleModel          697
ManufacturingYear     695
Type                  685
rating                690
color                2124
Duty                    0
fuel                  696
CylinderCount           0
type of gear          692
capacity                0
Odometer                0
#airbags                0
price                   0
dtype: int64

### Summary of test data

In [13]:
test.shape

(3000, 14)

In [14]:
test.describe()

ID       rating
count   3000.000000  2690.000000
mean    9149.806667     2.593309
std     5137.688945     1.891348
min        8.000000    -4.000000
25%     4804.500000     1.000000
50%     9126.000000     3.000000
75%    13553.000000     4.000000
max    18016.000000    10.000000

In [15]:
test.describe(include=['O'])

Brand VehicleModel ManufacturingYear   Type      color  Duty  fuel  \
count      2706         2697              2695   2685       2124  3000  2696   
unique      157          920               351     72         16   811     6   
top     hyundai        Prius              2014  sedan  Jet Black     -   Gas   
freq        212           49               180    492        555   916  1424   

       CylinderCount type of gear capacity Odometer #airbags  
count           3000         2692     3000     3000     3000  
unique            23            7       76     2411      243  
top              4.0         auto        2     0 km       12  
freq            1130         1215      529       44      356

In [16]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ID                 3000 non-null   int64  
 1   Brand              2706 non-null   object 
 2   VehicleModel       2697 non-null   object 
 3   ManufacturingYear  2695 non-null   object 
 4   Type               2685 non-null   object 
 5   rating             2690 non-null   float64
 6   color              2124 non-null   object 
 7   Duty               3000 non-null   object 
 8   fuel               2696 non-null   object 
 9   CylinderCount      3000 non-null   object 
 10  type of gear       2692 non-null   object 
 11  capacity           3000 non-null   object 
 12  Odometer           3000 non-null   object 
 13  #airbags           3000 non-null   object 
dtypes: float64(1), int64(1), object(12)
memory usage: 328.2+ KB


In [17]:
test.isnull().sum()

ID                     0
Brand                294
VehicleModel         303
ManufacturingYear    305
Type                 315
rating               310
color                876
Duty                   0
fuel                 304
CylinderCount          0
type of gear         308
capacity               0
Odometer               0
#airbags               0
dtype: int64

Let's start with checking each column and its data type.

## CylinderCount column

I started with this column because I saw that it doesn't have any missing values, so I thought it would help a lot in imputing missing values in other columns.

In [18]:
manipulated.CylinderCount.value_counts(dropna = False)

4.0         2638
4.0 cyl     1631
0            718
6.0          627
14.0         525
6.0 cyl      344
8.0          161
16.0         107
8.0 cyl       94
5.0           29
18.0          26
5.0 cyl       19
3.0           16
3.0 cyl       12
12.0           8
2.0            7
1.0            7
15.0           6
2.0 cyl        6
12.0 cyl       4
1.0 cyl        3
10.0           3
16.0 cyl       2
11.0           2
7.0            2
10.0 cyl       1
13.0           1
9.0            1
Name: CylinderCount, dtype: int64

In [19]:
def manipulate_cylindercount(data):
    
    # extract numeric values, and here I used regex to make it easy to capture the pattern
    data['CylinderCount'] = data['CylinderCount'].str.extract(r'(\d+)', expand=False)
    
    # convert column to numeric (integer)
    data['CylinderCount'] = pd.to_numeric(data['CylinderCount'])
    
    # convert column to a categorical data type
    data['CylinderCount'] = data['CylinderCount'].astype('category')
    
    return data


In [20]:
manipulated = manipulate_cylindercount(manipulated)

manipulated['CylinderCount'].isna().sum()

0

In [21]:
manipulated.CylinderCount.value_counts()

4     4269
6      971
0      718
14     525
8      255
16     109
5       48
3       28
18      26
2       13
12      12
1       10
15       6
10       4
11       2
7        2
13       1
9        1
Name: CylinderCount, dtype: int64

## Capacity column

In [22]:
manipulated['capacity'].value_counts(dropna = False)

2            1261
2.5           765
0             714
1.8           609
1.6           503
1.5           448
3.5           361
2.4           349
3             270
1.3           196
2.0 Turbo     122
1.4           121
4.6            93
3.3            78
2.2 Turbo      76
1.8 Turbo      63
3.0 Turbo      62
2.2            59
3.2            59
1.7            57
1.4 Turbo      49
1.6 Turbo      47
2.7            39
4.7            39
1.2            37
3.6            35
1              34
2.5 Turbo      31
2.3            29
4.4            29
4              29
1.5 Turbo      26
3.7            25
5              23
2.4 Turbo      19
5.7            19
5.5            18
2.7 Turbo      17
1.9            14
2.8            13
1.9 Turbo      11
1.7 Turbo      10
4.8             9
1.3 Turbo       8
0.4             8
2.6             7
3.5 Turbo       7
4.2             6
0.7             6
2.8 Turbo       6
6.2             6
3.2 Turbo       5
3.8             5
4.3             4
2.9 Turbo       4
4.5       

In [23]:
is_nan(manipulated['capacity'])

0

In [24]:
manipulated['capacity'] = manipulated['capacity'].astype('category')

I didn't do too much in this column as I thought it would be a good idea to keep things all as it is.

## Brand column

In [25]:
train['Brand'].unique()

array([nan, 'I love TOYOTA', 'hyundai', 'opel', 'subaru', 'VOLKSWAGEN',
       'FORD', 'Bmw', 'jaguar', 'I love HONDA', 'mercedes-benz',
       'I love FORD', 'honda', 'HYUNDAI', 'NISSAN', 'toyota',
       'volkswagen', 'lexus', 'ford', 'SSANGYONG', 'HONDA', 'Hyundai',
       'Toyota', 'BMW', 'MERCEDES-BENZ is the best',
       'I love MERCEDES-BENZ', 'SSANGYONG is the best', 'I love SUBARU',
       'ssangyong', 'TOYOTA is the best', 'nissan', 'kia', 'DAEWOO',
       'SUBARU is the best', 'TOYOTA', 'bmw', 'AUDI', 'CHEVROLET', 'jeep',
       'SUBARU', 'MERCEDES-BENZ', 'Mercedes-Benz', 'I love BMW', 'LEXUS',
       'I love HYUNDAI', 'NISSAN is the best', 'Opel', 'PORSCHE', 'MAZDA',
       'I love NISSAN', 'HYUNDAI is the best', 'chevrolet', 'mitsubishi',
       'I love KIA', 'Chevrolet', 'mazda', 'KIA', 'BMW is the best',
       'gmc', 'I love LEXUS', 'MITSUBISHI', 'I love MITSUBISHI', 'skoda',
       'Lexus', 'Volkswagen', 'SUZUKI', 'Ford', 'JEEP', 'OPEL', 'Honda',
       'I love VOLKSW

In [26]:
print(train['Brand'].nunique())

train['Brand'] = train['Brand'].str.lower()

print(train['Brand'].nunique())

185
114


In [27]:

def manipulate_brand(data):
    
    data['Brand'] = data['Brand'].str.lower()
    data['Brand'] = data['Brand'].str.replace('i love ', '')
    data['Brand'] = data['Brand'].str.replace(' is the best', '')
    
    # I searched the model of this row and I found that its brand was GAC, it won't affect anyway but I just did it :)
    data.loc[data['Brand'] == 'სხვა', 'Brand'] = 'gac' 
    
    return data

In [28]:
manipulated = manipulate_brand(manipulated)
manipulated['Brand'].unique()

array(['nan', 'toyota', 'hyundai', 'opel', 'subaru', 'volkswagen', 'ford',
       'bmw', 'jaguar', 'honda', 'mercedes-benz', 'nissan', 'lexus',
       'ssangyong', 'kia', 'daewoo', 'audi', 'chevrolet', 'jeep',
       'porsche', 'mazda', 'mitsubishi', 'gmc', 'skoda', 'suzuki', 'fiat',
       'peugeot', 'dodge', 'mini', 'vaz', 'gaz', 'cadillac', 'uaz',
       'seat', 'rover', 'chrysler', 'land rover', 'infiniti', 'gac',
       'alfa romeo', 'renault', 'greatwall', 'buick', 'volvo', 'lincoln',
       'isuzu', 'scion', 'rolls-royce', 'citroen', 'pontiac', 'acura',
       'daihatsu', 'moskvich', 'mercury', 'maserati'], dtype=object)

In [29]:
manipulated.Brand.nunique()

55

In [30]:
manipulated.Brand.value_counts(dropna = False)


toyota           1246
hyundai          1225
nan               706
mercedes-benz     651
chevrolet         362
ford              353
bmw               333
honda             325
lexus             279
nissan            216
volkswagen        194
ssangyong         163
kia               155
opel              139
subaru            105
mitsubishi         96
audi               78
mazda              61
jeep               44
fiat               30
dodge              30
daewoo             30
suzuki             24
renault            13
jaguar             12
porsche            11
mini               11
vaz                11
lincoln            10
land rover          9
skoda               7
peugeot             7
volvo               7
infiniti            7
buick               6
daihatsu            5
uaz                 5
gaz                 4
chrysler            3
citroen             3
scion               3
cadillac            3
acura               2
moskvich            2
alfa romeo          2
gmc       

We have 706 missing values in the brand column, so we will move on exploring the dataset more to figure out how to fill these missing values.

## VehicleModel column

In [31]:
train["VehicleModel"].nunique()

1480

In [32]:
train["VehicleModel"].value_counts(dropna = False)

NaN                                                                                   697
Prius                                                                                 113
Sonata                                                                                104
Elantra                                                                                96
FIT                                                                                    96
PRIUS                                                                                  89
H1                                                                                     87
Camry                                                                                  84
E 350                                                                                  73
SONATA                                                                                 72
I have the fastest model ever.. yes you got it right it's Sonata                       68
ELANTRA   

This part was just to see how many rows with weired words (Georgian words as I figured it later)

In [33]:
first = train["VehicleModel"].value_counts(dropna = False)

# this regex will remove all non-ascii characters from (0 - 127) and replace them with an empty string
second = train["VehicleModel"].str.replace(r'[^\x00-\x7F]', '', regex=True).value_counts(dropna = False)

In [34]:
count = 0
for i in first.index:
    if i not in second.index:
        count += 1

count

11

- So we have 11 Georigian letters in the VehicleModel column we don't need, so we will remove them.

In [35]:
manipulated["VehicleModel"] = manipulated["VehicleModel"].str.replace(r'[^\x00-\x7F]', '', regex=True).str.strip()

In [36]:
manipulated["VehicleModel"].nunique()

1477

In [37]:
manipulated["VehicleModel"].value_counts(dropna = False)

nan                                                                                   697
Prius                                                                                 113
Sonata                                                                                104
FIT                                                                                    96
Elantra                                                                                96
PRIUS                                                                                  89
H1                                                                                     87
Camry                                                                                  84
E 350                                                                                  74
SONATA                                                                                 72
I have the fastest model ever.. yes you got it right it's Sonata                       68
ELANTRA   

In [38]:
def manipulate_vehiclemodel(data):
    
    # to lower case all to ease the manipulation
    data["VehicleModel"] = data["VehicleModel"].str.lower()
    
    # to remove some unnecessary words
    to_remove = ["i have the fastest model ever.. yes you got it right it's ", "my model which is ", " is the best", "i own " ]
    for i in to_remove:
        data["VehicleModel"] = data["VehicleModel"].str.replace(i, "")
    
    # for removing trailing dots
    data["VehicleModel"] = data["VehicleModel"].str.replace('.', "")
    
    # for removing non-ascii characters
    data["VehicleModel"] = data["VehicleModel"].str.replace(r'[^\x00-\x7F]', '', regex=True).str.strip()
    
    
    """
    This part was to make the model just one word for each group of brand, for example if the model was "prius v" it will be just "prius"
    so by that way we can reduce the number of unique values in the model column and make it more consistent as I thought.
    """
    brands = data['Brand'].unique()
    for brand in brands:
        mask = data['Brand'] == brand
        data.loc[mask, 'VehicleModel'] = data.loc[mask, 'VehicleModel'].astype(str)
        
        # process 'VehicleModel' column to keep only the first word if it contains more than one word
        data.loc[mask, 'VehicleModel'] = data.loc[mask, 'VehicleModel'].apply(lambda x: \
                                                                            ''.join(str(x).split(" ")[0]) if len(str(x).split(" ")) > 1 \
                                                                                                            else x)
    return data

In [39]:
manipulated = manipulate_vehiclemodel(manipulated)
manipulated.VehicleModel.nunique()

430

In [40]:
manipulated.VehicleModel.value_counts()

nan           697
prius         437
sonata        369
camry         341
elantra       329
e             255
fit           172
h1            157
tucson        148
santa         137
aqua          124
x5            123
cruze         111
fusion         99
gx             98
ml             94
optima         91
transit        91
rx             90
highlander     87
jetta          82
c              81
actyon         73
rav            64
rexton         63
juke           61
astra          61
escape         54
forester       51
volt           50
insight        48
civic          46
lacetti        43
ct             43
captiva        42
passat         39
orlando        37
pajero         36
focus          35
malibu         35
vito           32
chr            31
a4             29
corolla        29
golf           28
sprinter       27
328            26
genesis        26
vectra         26
korando        25
tacoma         25
gla            24
320            23
tiida          23
500            22
outlander 

Now we noticed that it dropped to 430 after 1477 unique values and we didn't lose so much information!

In [41]:
manipulated["VehicleModel"].nunique()

430

In [42]:
is_nan(manipulated['VehicleModel'])

697

Now let's try to fill the missing values in the model column.
- We will do that by following these steps:
    1. First we will replace 'nan' strings values with real NaN values to fill it correctly.
    2. Second we will group the data by brand, cylinder count, and capacity as I saw that model of the car depends on these feature the most.
    3. Third we will fill the missing values in the model column by the most frequent value in the group.
    4. Finally, we will fill the remaining missing values with 'unknown' category.

In [43]:
def fill_missing_model(data):
    
    data['VehicleModel'] = data['VehicleModel'].replace('nan', pd.NA)
    to_fill_model = data.groupby(['Brand', 'CylinderCount', 'capacity'])['VehicleModel']
    data['VehicleModel'] = to_fill_model.transform(lambda x: x.fillna(x.mode().iloc[0] if not x.mode().empty else 'unknown'))
    return data


In [44]:
manipulated = fill_missing_model(manipulated)

manipulated['VehicleModel'].isna().sum()

0

In [45]:
is_nan(manipulated['Brand'])

706

For the brand column, we will follow the same technique as the model column.

But I just grouped by the model as I saw that the brand of the car depends just on the model.

In [46]:

def fill_missing_brand(data):
    data['Brand'] = data['Brand'].replace('nan', pd.NA)
    to_fill_brands = data.groupby('VehicleModel')['Brand']
    data['Brand'] = to_fill_brands.transform(lambda x: x.fillna(x.mode().iloc[0] if not x.mode().empty else 'unknown'))
    return data

In [47]:
manipulated = fill_missing_brand(manipulated)

In [48]:
print(is_nan(manipulated['Brand']))

print(is_nan(manipulated['VehicleModel']))


0
0


## ManufacturingYear column

In [49]:
is_nan(train["ManufacturingYear"])

695

In [50]:
import regex as re
s = "2004%"
re.findall(r'\d{4}', s)

['2004']

In [51]:

# Extract the 4-digit year from 'ManufacturingYear' column
reg_year = train["ManufacturingYear"].str.extract(r'(\d{4})', expand=False)
    
# Extract any numeric value followed by 'years ago' from 'ManufacturingYear' column
reg_old = train["ManufacturingYear"].str.extract(r'(\d+)y', expand=False)



# I just did these steps to make sure that I extracted the correct values and didn't make any nan values
print(len(reg_year.dropna()))
print(reg_year.dropna().unique())
print()
print(len(reg_old.dropna()))
print(reg_old.dropna().unique())

nan_values = len(train) - (len(reg_old.dropna()) + len(reg_year.dropna()))

print()
print("Nan values equals to: ", nan_values)

5669
['2017' '2009' '2007' '3010' '1999' '2012' '2011' '2014' '2015' '2006'
 '2105' '2013' '2010' '2003' '2016' '1998' '2018' '2008' '2111' '2020'
 '2005' '2004' '1910' '1914' '3009' '2001' '2112' '1997' '1917' '1916'
 '1912' '2002' '2110' '2115' '2019' '3013' '1907' '2000' '1996' '3014'
 '3004' '1913' '1900' '1902' '2117' '2118' '1995' '2113' '3012' '1915'
 '1911' '2116' '2098' '1918' '2114' '3008' '3006' '2107' '2119' '1908'
 '1992' '2099' '1985' '1909' '3007' '2103' '1905' '1983' '3011' '2104'
 '3015' '1904' '1993' '1994' '1988' '3017' '1898' '1903' '3018' '2998'
 '3016' '1982' '2108' '2106' '3003' '1896' '2100' '1893' '1987' '1991'
 '2120' '1990' '2102' '3005' '2097' '1939' '1919' '2096' '2101' '2999'
 '1901' '3019' '3002' '1986' '3000' '2109' '1920' '2089' '2992' '3001'
 '1984' '1906' '2093' '1978' '2997' '1895']

636
['14' '15' '10' '12' '13' '11' '7' '8' '6' '9' '21' '26' '28' '24' '20'
 '23' '33' '17' '25' '19' '16' '29' '18' '5' '22' '31' '4' '27' '71' '40']

Nan values equals

Now let's do the magic:

In [52]:
# this might not be the best way to do it but I didn't want to waste more time in this column
# this function will make the year reasonable and consistent

def check_years(year):
    if year == str(pd.NA):
        return year
    if year[0] == '3':
        return '2'+ year[1:]
    if year[0] == '2' and year[1] >='8': 
        return '19'+ year[2:]
    if year[0] == '2' and year[1] != '0':
        return '20'+ year[2:]
    elif year[0] == '2' and year[1] == '0':
        if year[2] <= '2':
            return '20'+ year[2:]
        else:
            return '19'+ year[2:]
    return year

In [53]:

def manipulate_years(data):
    
    def extract_year(year_str):
        
        import re
        match = re.search(r'(\d{4})', year_str)
        if match:
            return match.group(1)
        else:
            match = re.search(r'(\d+)y', year_str)
            if match:
                return 2021 - int(match.group(1))
            else:
                return pd.NA


    data["ManufacturingYear"] = data["ManufacturingYear"].astype(str).apply(extract_year)
    data.ManufacturingYear = data.ManufacturingYear.apply(lambda x: check_years(str(x)))
    data.ManufacturingYear = pd.to_numeric(data.ManufacturingYear, errors='coerce')
    return data


In [54]:
manipulated = manipulate_years(manipulated)

print(is_nan(manipulated.ManufacturingYear))

print(manipulated.ManufacturingYear.nunique())

manipulated.ManufacturingYear.unique()

695
68


array([2017.,   nan, 2009., 2007., 2010., 1999., 2012., 2011., 2014.,
       2015., 2006., 2005., 2013., 2003., 2016., 1998., 2018., 2008.,
       2020., 2004., 1910., 1914., 2001., 1997., 1917., 1916., 1912.,
       2002., 2019., 1907., 2000., 1996., 1913., 1900., 1902., 1995.,
       1915., 1911., 1918., 1993., 1908., 1992., 1985., 1909., 1905.,
       1988., 1983., 1904., 1994., 1898., 1903., 1982., 1896., 1893.,
       1987., 1991., 1990., 1939., 1919., 1901., 1950., 1986., 1981.,
       1920., 1989., 1984., 1906., 1978., 1895.])

To fill missing values in the manufacturing year column, we will do the same steps as we did in the model and brand columns.

In [55]:
def fill_missing_year(data):
    # Group by 'Brand' and 'VehicleModel'
    to_fill_years = data.groupby(['Brand', 'VehicleModel'])['ManufacturingYear']
    
    # Fill missing values in 'ManufacturingYear' with the most frequent year within the group,
    # or the overall mode if the group mode is empty
    data['ManufacturingYear'] = to_fill_years.transform(
        lambda x: x.fillna(
            x.mode().iloc[0] if not x.mode().empty else data['ManufacturingYear'].mode().iloc[0]
        )
    )
    
    # Convert 'ManufacturingYear' to integer type
    data['ManufacturingYear'] = data['ManufacturingYear'].astype(int)
    
    # Convert 'ManufacturingYear' to datetime and then to period (year only)
    # this is not a necessary step but I just did it to make the data more consistent
    data['ManufacturingYear'] = pd.to_datetime(data['ManufacturingYear'], format='%Y').dt.to_period('Y')
    
    return data

In [56]:

manipulated = fill_missing_year(manipulated)

print(manipulated['ManufacturingYear'].isna().sum())

0


In [57]:
manipulated.ManufacturingYear.value_counts().sort_index()

1893      1
1895      1
1896      3
1898      4
1900      5
1901      2
1902      8
1903      2
1904      6
1905      5
1906      2
1907      7
1908     10
1909      6
1910     21
1911     28
1912     24
1913     23
1914     21
1915     11
1916     15
1917      9
1918      7
1919      6
1920      1
1939      1
1950      1
1978      1
1981      1
1982      2
1983      4
1984      1
1985      2
1986      1
1987      2
1988      4
1989      1
1990      8
1991      3
1992     23
1993     10
1994     15
1995     39
1996     32
1997     55
1998     79
1999     68
2000    118
2001     86
2002    111
2003    148
2004    128
2005    141
2006    132
2007    217
2008    295
2009    254
2010    590
2011    659
2012    743
2013    685
2014    748
2015    527
2016    415
2017    239
2018    111
2019     64
2020      8
Freq: A-DEC, Name: ManufacturingYear, dtype: int64

Now we are good to go with the next column.

## Type column

In [58]:
train['Type'].value_counts(dropna = False)

sedan               1160
SEDAN                850
jeep                 691
NaN                  685
JEEP                 501
hatchback            383
Sedan                302
HATCHBACK            285
WOOW Sedan           270
Jeep                 175
WOOW Jeep            160
Hatchback            117
minivan              103
WOOW Hatchback        96
MINIVAN               82
Sedan0                73
Sedan2                67
Sedan4                67
Sedan3                64
coupe                 54
Sedan1                54
COUPE                 48
universal             45
Jeep1                 44
Jeep0                 40
goods wagon           39
Jeep3                 39
microbus              36
UNIVERSAL             34
Jeep2                 33
MICROBUS              33
Hatchback3            33
Minivan               27
GOODS WAGON           24
Jeep4                 23
Hatchback4            21
WOOW Coupe            20
Hatchback0            19
Hatchback1            18
Hatchback2            17


In [59]:
def manipulate_type(data):

    data['Type'] = data['Type'].str.lower()  # Convert to lowercase
    data['Type'] = data['Type'].str.replace('woow ', '')  # Remove 'woow '
    data['Type'] = data['Type'].str.replace(r'\d+', '', regex=True)  # Remove digits

    return data


In [60]:
manipulated = manipulate_type(manipulated)

manipulated['Type'].value_counts(dropna = False)

sedan          2907
jeep           1706
hatchback       989
nan             685
minivan         240
coupe           153
universal       107
microbus        101
goods wagon      87
pickup           14
cabriolet         9
limousine         2
Name: Type, dtype: int64

In [61]:
print (is_nan(manipulated['Type']))

685


The same here for the type column, we will do the same steps as we did in the model, brand, and manufacturing year columns.

In [62]:

def fill_missing_types(data):
    data['Type'] = data['Type'].replace('nan', pd.NA)
    to_fill_types = data.groupby(['Brand', 'VehicleModel'])['Type']
    data['Type'] = to_fill_types.transform(lambda x: x.fillna(x.mode().iloc[0] if not x.mode().empty else 'unknown'))
    return data

In [63]:
manipulated = fill_missing_types(manipulated)

is_nan(manipulated['Type'])

0

In [64]:
manipulated['Type'].value_counts(dropna = False)


sedan          3179
jeep           1905
hatchback      1129
minivan         264
coupe           166
microbus        112
universal       108
goods wagon      89
unknown          22
pickup           15
cabriolet         9
limousine         2
Name: Type, dtype: int64

## Rating column

In [65]:
train['rating'].value_counts(dropna = False)

 2.0     1248
 3.0     1175
 1.0      935
 4.0      925
 0.0      857
 NaN      690
 5.0      609
 6.0      274
-1.0      108
 7.0       94
 8.0       35
-2.0       30
-3.0       11
 9.0        7
 10.0       1
-4.0        1
Name: rating, dtype: int64

In [66]:
manipulated['rating'].isna().sum()

690

The same steps also to impute missing values:

In [67]:

def fill_missing_ratings(data):
    to_fill_ratings = data.groupby(['Brand', 'VehicleModel', 'Type'])['rating']
    data['rating'] = to_fill_ratings.transform(lambda x: x.fillna(x.mode().iloc[0] if not x.mode().empty else 0))
    data['rating'] = data['rating'].astype('category')
    return data

In [68]:
manipulated = fill_missing_ratings(manipulated)

manipulated['rating'].isna().sum()

0

In [69]:
manipulated['rating'].value_counts(dropna = False)

2.0     1539
3.0     1351
1.0      997
0.0      977
4.0      949
5.0      621
6.0      275
-1.0     109
7.0       95
8.0       35
-2.0      32
-3.0      11
9.0        7
-4.0       1
10.0       1
Name: rating, dtype: int64

## Color column

In [70]:
train['color'].value_counts(dropna = False)

NaN              2124
Jet Black        1239
Pearl White      1083
Metallic          995
Charcoal          628
Sapphire Blue     371
Crimson           166
Emerald Green      77
Tangerine          66
Gold               49
Chocolate          47
Ruby Red           44
Sky Blue           35
Sand               34
Lemon Yellow       30
Royal Purple        7
Rose Pink           5
Name: color, dtype: int64

In [71]:
manipulated.color.value_counts(dropna=False)

nan              2124
Jet Black        1239
Pearl White      1083
Metallic          995
Charcoal          628
Sapphire Blue     371
Crimson           166
Emerald Green      77
Tangerine          66
Gold               49
Chocolate          47
Ruby Red           44
Sky Blue           35
Sand               34
Lemon Yellow       30
Royal Purple        7
Rose Pink           5
Name: color, dtype: int64

In [72]:
is_nan(manipulated['color'])

2124

In [73]:
import pandas as pd

def fill_manipulate_missing_color(data):
    # Convert color to lowercase
    data['color'] = data['color'].str.lower()
    
    # Replace 'nan' with actual missing values
    data['color'].replace('nan', pd.NA, inplace=True)
    
    # Group by 'Brand', 'VehicleModel', and 'Type' and calculate the mode for 'color'
    to_fill_color = data.groupby(['Brand', 'VehicleModel', 'Type'])['color']
    
    # Fill missing values with the mode within each group
    data['color'] = to_fill_color.transform(lambda x: x.fillna(x.mode().iloc[0] if not x.mode().empty else 'unknown'))
    
    return data

In [74]:
manipulated = fill_manipulate_missing_color(manipulated)

manipulated['color'].value_counts(dropna = False)

jet black        1978
metallic         1594
pearl white      1514
charcoal          790
sapphire blue     397
crimson           186
unknown           131
emerald green      81
tangerine          66
gold               57
chocolate          49
ruby red           44
sand               35
sky blue           35
lemon yellow       31
royal purple        7
rose pink           5
Name: color, dtype: int64

In [75]:
manipulated['color'] = manipulated['color'].astype('category')

## Duty column

In [76]:
train['Duty'].value_counts(dropna = False)

-                                                                2108
640                                                                48
765                                                                45
891                                                                44
765 pound                                                          43
640$                                                               41
779                                                                40
765$                                                               37
781                                                                35
639 pound                                                          34
777                                                                33
707                                                                33
640 pound                                                          32
642                                                                32
891$                

In [77]:
def manipulate_duty(data):
    data['Duty'] = data['Duty'].str.extract(r'(\d+)', expand=False)
    data['Duty'] = pd.to_numeric(data['Duty'], errors='coerce')
    return data

In [78]:
manipulated = manipulate_duty(manipulated)

manipulated['Duty'].isna().sum()

2578

In [79]:
def fill_missing_duty(data):
    to_fill = data.groupby('Brand')['Duty'].mean().fillna(0).to_dict()
    data['Duty'] = data['Duty'].fillna(data['Brand'].map(to_fill))
    data['Duty'] = data['Duty'].astype(int)
    return data

In [80]:
manipulated = fill_missing_duty(manipulated)

is_nan(manipulated['Duty'])

0

In [81]:
manipulated.head()

ID    Brand VehicleModel ManufacturingYear       Type rating  \
0   2624    mazda        mazda              2017      sedan    0.0   
1  14392   toyota        camry              2014    minivan    5.0   
2   9822  hyundai           h1              2009    minivan    0.0   
3  17123     opel        corsa              2003  hatchback    5.0   
4  11133   subaru      outback              2007  hatchback    4.0   

           color  Duty  fuel CylinderCount type of gear capacity  \
0       charcoal  1249   Gas             4            R      2.5   
1       metallic  1754   Gas             4            R      2.4   
2       metallic  1185  Dies             4            R      2.5   
3      jet black  2076   nan             4         auto        0   
4  sapphire blue  2062   Gas             4          nan      2.5   

         Odometer #airbags  price  
0        75492 KM     1008  21012  
1       220000 KM      -96  11917  
2  TOTAL 50653 km  4 count  22433  
3       230000 KM      0     5645  
4       131000 KM      4    10976

## Fuel column

In [82]:
manipulated.fuel.value_counts(dropna = False)

Gas                3272
Dies               1291
Electric Hybrid    1250
nan                 696
other               307
Natural Gas         157
hyb                  26
others                1
Name: fuel, dtype: int64

In [83]:
def fill_manipulate_fuel(data):
    # Convert fuel to lowercase
    data['fuel'] = data['fuel'].str.lower()
    
    # Replace 'nan' with actual missing values
    data['fuel'].replace('nan', pd.NA, inplace=True)
    
    # Group by 'Brand', 'VehicleModel', and 'Type' and calculate the mode for 'fuel'
    to_fill_fuel = data.groupby(['Brand', 'VehicleModel', 'Type'])['fuel']
    
    # Fill missing values with the mode within each group
    data['fuel'] = to_fill_fuel.transform(lambda x: x.fillna(x.mode().iloc[0] if not x.mode().empty else 'other'))
    data['fuel'] = data['fuel'].astype('category')
    
    return data

In [84]:
manipulated = fill_manipulate_fuel(manipulated)

manipulated['fuel'].value_counts(dropna = False)

gas                3563
dies               1444
electric hybrid    1415
other               383
natural gas         165
hyb                  29
others                1
Name: fuel, dtype: int64

## Type of gear column

In [85]:
manipulated['type of gear'].value_counts(dropna = False)

auto      2870
others    1228
nan        692
gear       639
R          635
AUTO       553
manual     320
MANUAL      63
Name: type of gear, dtype: int64

In [86]:
def fill_manipulate_type_of_gear(data):
    
    data['type of gear'] = data['type of gear'].str.lower()
    data['type of gear'].replace('nan', pd.NA, inplace = True)
    to_fill_gear = data.groupby(['Brand', 'VehicleModel', 'Type'])['type of gear']
    data['type of gear'] = to_fill_gear.transform(lambda x: x.fillna(x.mode().iloc[0] if not x.mode().empty else 'others'))
    data['type of gear'] = data['type of gear'].astype('category')
    
    return data
    

In [87]:
manipulated = fill_manipulate_type_of_gear(manipulated)

manipulated['type of gear'].value_counts(dropna = False)

auto      4011
others    1273
gear       649
r          646
manual     421
Name: type of gear, dtype: int64

## Odometer column

In [88]:
manipulated['Odometer'].value_counts(dropna = False)

0 km                   79
0  kilo                52
0 KM                   48
0                      30
TOTAL 0 km             30
150000 km              19
180000 km              18
0  kilometer           18
150000 KM              15
200000 KM              14
170000 km              14
120000 km              14
200000  kilo           14
100000 km              13
100000  kilo           12
145000 km              12
200000 km              12
160000 km              11
190000 km              11
111111 km              11
1000  kilo             10
TOTAL 200000 km        10
100000 KM              10
250000 km              10
300000 km               9
200000                  9
1000 KM                 9
160000  kilo            9
127000 km               9
150000  kilo            9
160000 KM               9
260000 km               8
105000 km               8
11111  kilo             8
110000 km               8
190000  kilo            8
245000 km               8
220000  kilo            8
180000  kilo

In [89]:
def manipulate_odometer(data):
    data['Odometer'] = data['Odometer'].str.replace(r'\D', '', regex=True)
    data['Odometer'] = pd.to_numeric(data['Odometer'], errors='coerce')
    data['Odometer'].rename('Odometer_km', inplace=True)
    return data

In [90]:
manipulated = manipulate_odometer(manipulated)

manipulated['Odometer'].isna().sum()

0

In [91]:
manipulated['Odometer'].value_counts(dropna = False).sort_index()

0             257
18              1
120             1
261             1
429             1
498             1
500             3
554             1
555             1
777             1
1000           34
1111            7
1200            1
1212            1
1234            2
1300            2
1405            1
1461            1
1500            1
1544            1
1600            6
1669            2
1717            1
1792            1
1800            1
2000            6
2014            1
2034            1
2102            1
2336            1
2342            1
2598            1
2635            1
2642            1
2700            1
2741            1
2888            1
3000            3
3100            1
3200            2
3474            1
3500            1
3854            1
4000            1
4088            1
4130            1
4330            1
4363            1
4444            1
4512            1
4911            1
5000            3
5700            1
5987            1
6000            1
6384      

## Airbags column

In [92]:
manipulated.rename(columns = {'#airbags':'airbags'}, inplace=True)

In [93]:
manipulated['airbags'].value_counts(dropna = False)

12                830
4                 828
0                 349
8                 248
4                 233
6                 219
ONLY 12 AIRBAG    214
12 bags           210
4 bags            210
ONLY 4 AIRBAG     207
12                196
4 count           187
12 count          180
2                 165
10                143
0                  88
0 count            87
0 bags             79
-96                78
1012               77
ONLY 0 AIRBAG      72
8 count            71
-88                69
ONLY 8 AIRBAG      65
104                60
6                  58
112                57
6 bags             56
8 bags             55
1004               52
8                  51
ONLY 6 AIRBAG      45
2 count            44
2                  40
ONLY 2 AIRBAG      40
2 bags             36
-100               36
10 bags            32
6 count            30
10                 25
10 count           25
1000               25
ONLY 10 AIRBAG     24
1008               24
108                23
100       

In [94]:
def manipulate_airbags(data):
    
    data['airbags'] = data['airbags'].str.extract(r'(-?\d+)', expand=False)
    data['airbags'] = pd.to_numeric(data['airbags'], errors='coerce')
    
    for i in data['airbags'].unique():
        if i < 0:
            data['airbags'] = data['airbags'].replace(i, 0)
        if i > 12:
            data['airbags'] = data['airbags'].replace(i, 12)
    data['airbags'] = data['airbags'].astype(int)
    
    return data

In [95]:
manipulated = manipulate_airbags(manipulated)

manipulated['airbags'].value_counts(dropna = False).sort_index()

0     1029
1       33
2      357
3       11
4     1870
5       34
6      455
7       30
8      545
9       19
10     270
11      16
12    2331
Name: airbags, dtype: int64

In [96]:
manipulated['airbags'].isna().sum()

0

In [97]:
manipulated.shape

(7000, 15)

In [98]:
manipulated.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7000 entries, 0 to 6999
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype        
---  ------             --------------  -----        
 0   ID                 7000 non-null   int64        
 1   Brand              7000 non-null   object       
 2   VehicleModel       7000 non-null   object       
 3   ManufacturingYear  7000 non-null   period[A-DEC]
 4   Type               7000 non-null   object       
 5   rating             7000 non-null   category     
 6   color              7000 non-null   category     
 7   Duty               7000 non-null   int32        
 8   fuel               7000 non-null   category     
 9   CylinderCount      7000 non-null   category     
 10  type of gear       7000 non-null   category     
 11  capacity           7000 non-null   category     
 12  Odometer           7000 non-null   int64        
 13  airbags            7000 non-null   int32        
 14  price              7000 

- Now let's do the same for the test data.

## Test data

In [99]:
test.shape

(3000, 14)

In [100]:
for col in test.columns:
    print(f"{col}: {test[col].isna().sum()}")

ID: 0
Brand: 294
VehicleModel: 303
ManufacturingYear: 305
Type: 315
rating: 310
color: 876
Duty: 0
fuel: 304
CylinderCount: 0
type of gear: 308
capacity: 0
Odometer: 0
#airbags: 0


In [101]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ID                 3000 non-null   int64  
 1   Brand              2706 non-null   object 
 2   VehicleModel       2697 non-null   object 
 3   ManufacturingYear  2695 non-null   object 
 4   Type               2685 non-null   object 
 5   rating             2690 non-null   float64
 6   color              2124 non-null   object 
 7   Duty               3000 non-null   object 
 8   fuel               2696 non-null   object 
 9   CylinderCount      3000 non-null   object 
 10  type of gear       2692 non-null   object 
 11  capacity           3000 non-null   object 
 12  Odometer           3000 non-null   object 
 13  #airbags           3000 non-null   object 
dtypes: float64(1), int64(1), object(12)
memory usage: 328.2+ KB


In [102]:
for col in test.columns[test.dtypes == 'object']:
    test[col] = test[col].astype('str')

In [103]:
test = manipulate_cylindercount(test)
test['capacity'] = test['capacity'].astype('category')
test = manipulate_brand(test)
test = manipulate_vehiclemodel(test)
test = fill_missing_model(test)
test = fill_missing_brand(test)
test = manipulate_years(test)
test = fill_missing_year(test)
test = manipulate_type(test)
test = fill_missing_types(test)
test = fill_missing_ratings(test)
test = fill_manipulate_missing_color(test)
test = manipulate_duty(test)
test = fill_missing_duty(test)
test = fill_manipulate_fuel(test)
test = fill_manipulate_type_of_gear(test)
test = manipulate_odometer(test)
test.rename(columns = {'#airbags':'airbags'}, inplace=True)
test = manipulate_airbags(test)

In [104]:
for col in test.columns:
    print(f"{col}: {test[col].isna().sum()}")

ID: 0
Brand: 0
VehicleModel: 0
ManufacturingYear: 0
Type: 0
rating: 0
color: 0
Duty: 0
fuel: 0
CylinderCount: 0
type of gear: 0
capacity: 0
Odometer: 0
airbags: 0


In [105]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype        
---  ------             --------------  -----        
 0   ID                 3000 non-null   int64        
 1   Brand              3000 non-null   object       
 2   VehicleModel       3000 non-null   object       
 3   ManufacturingYear  3000 non-null   period[A-DEC]
 4   Type               3000 non-null   object       
 5   rating             3000 non-null   category     
 6   color              3000 non-null   object       
 7   Duty               3000 non-null   int32        
 8   fuel               3000 non-null   category     
 9   CylinderCount      3000 non-null   category     
 10  type of gear       3000 non-null   category     
 11  capacity           3000 non-null   category     
 12  Odometer           3000 non-null   int64        
 13  airbags            3000 non-null   int32        
dtypes: category(5), int32(2)

In [106]:
test.shape

(3000, 14)

In [107]:
manipulated.to_csv('train_ready.csv', index=False)
test.to_csv('test_ready.csv', index=False)

Here we finished the cleaning process for the train data and test data, and we are ready to move on to the next step, which is preprocessing the data.

That's it for now, and we will continue in the next step.